pkg/vere/mars.c

```diff
@@ -0,0 +1,288 @@
+/* worker/mars.c
+**
+**  the main loop of a mars process.
+*/
+#include "c3.h"
+#include "noun.h"
+#include "types.h"
+#include "vere.h"
+#include "db/lmdb.h"
+#include <mars.h>
+#include <stdio.h>
+
+/* _mars_step_trace(): initialize or rotate trace file.
+*/
+static void
+_mars_step_trace(const c3_c* dir_c)
+{
+  if ( u3C.wag_w & u3o_trace ) {
+    c3_w trace_cnt_w = u3t_trace_cnt();
+    if ( trace_cnt_w == 0  && u3t_file_cnt() == 0 ) {
+      u3t_trace_open(dir_c);
+    }
+    else if ( trace_cnt_w >= 100000 ) {
+      u3t_trace_close();
+      u3t_trace_open(dir_c);
+    }
+  }
+}
+
+/* _mars_poke_play(): replay an event.
+*/
+static u3_weak
+_mars_poke_play(u3_mars* mar_u, c3_d eve_d, u3_noun job)
+{
+  u3_noun vir;
+
+  if ( c3n == u3v_poke_sure(0, job, &vir) ) {
+    return vir;
+  }
+
+  u3z(vir);
+  return u3_none;
+}
+
+typedef enum {
+  _play_yes_e,  //  success
+  _play_mem_e,  //  %meme
+  _play_int_e,  //  %intr
+  _play_log_e,  //  event log fail
+  _play_mug_e,  //  mug mismatch
+  _play_bad_e   //  total failure
+} _mars_play_e;
+
+/* _mars_play_batch(): replay a batch of events.
+*/
+static _mars_play_e
+_mars_play_batch(u3_mars* mar_u, c3_o mug_o, c3_w bat_w)
+{
+  u3_disk*      log_u = mar_u->log_u;
+  u3_disk_walk* wok_u = u3_disk_walk_init(log_u, mar_u->dun_d + 1, bat_w);
+  u3_fact       tac_u;
+  u3_noun         dud;
+
+  while ( c3y == u3_disk_walk_live(wok_u) ) {
+    if ( c3n == u3_disk_walk_step(wok_u, &tac_u) ) {
+      u3_disk_walk_done(wok_u);
+      return _play_log_e;
+    }
+
+    u3_assert( ++mar_u->sen_d == tac_u.eve_d );
+
+    if ( u3_none != (dud = _mars_poke_play(mar_u, tac_u.eve_d, tac_u.job)) ) {
+      c3_m mot_m;
+
+      mar_u->sen_d = mar_u->dun_d;
+      u3_disk_walk_done(wok_u);
+
+      u3_assert( c3y == u3r_safe_word(u3h(dud), &mot_m) );
+
+      switch ( mot_m ) {
+        case c3__meme: {
+          fprintf(stderr, "play (%" PRIu64 "): %%meme\r\n", tac_u.eve_d);
+          u3z(dud);
+          return _play_mem_e;
+        }
+
+        case c3__intr: {
+          fprintf(stderr, "play (%" PRIu64 "): %%intr\r\n", tac_u.eve_d);
+          u3z(dud);
+          return _play_int_e;
+        }
+
+        default: {
+          fprintf(stderr, "play (%" PRIu64 "): failed\r\n", tac_u.eve_d);
+          u3_pier_punt_goof("play", dud);
+          //  XX say something uplifting
+          //
+          return _play_bad_e;
+        }
+      }
+    }
+
+    mar_u->mug_l = u3r_mug(u3A->roc);
+
+    if ( tac_u.mug_l && (mar_u->mug_l != tac_u.mug_l) ) {
+      fprintf(stderr, "play (%" PRIu64 "): mug mismatch "
+                      "expected %08x, actual %08x\r\n",
+                      tac_u.eve_d, tac_u.mug_l, mar_u->mug_l);
+
+      if ( c3y == mug_o ) {
```

Comment:
Mug mismatch checks have been soft/advisory since mid-2020. They had been added and mandatory roughly a year prior to that. The problem was that they actually occurred with some frequency, and there was no way for people to work around / get past them.

They're actually quite rare now (as far as I know), and always indicate a serious problem. But people still need to be able to get past them in some scenarios (attempt to recover data, for instance). So I've made the checks hard by default, and added an argument to soften them up and continue replay.